runtime/src/main/kotlin/kotlin/collections/ArrayList.kt

```diff
@@ -9,52 +9,45 @@ actual class ArrayList<E> private constructor(
         private var array: Array<E>,
         private var offset: Int,
         private var length: Int,
+        private var isReadOnly: Boolean,
         private val backing: ArrayList<E>?
-) : MutableList<E>, RandomAccess, AbstractMutableCollection<E>() {
+) : MutableList<E>, RandomAccess, AbstractMutableList<E>() {
 
     actual constructor() : this(10)
 
     actual constructor(initialCapacity: Int) : this(
-            arrayOfUninitializedElements(initialCapacity), 0, 0, null)
+            arrayOfUninitializedElements(initialCapacity), 0, 0, false, null)
 
     actual constructor(elements: Collection<E>) : this(elements.size) {
         addAll(elements)
     }
 
-    override actual val size : Int
+    @PublishedApi
+    internal fun build(): List<E> {
+        if (backing != null) throw IllegalStateException() // just in case somebody casts subList to ArrayList
+        checkIsMutable()
+        isReadOnly = true
+        return this
+    }
+
+    override actual val size: Int
         get() = length
 
     override actual fun isEmpty(): Boolean = length == 0
 
     override actual fun get(index: Int): E {
-        checkIndex(index)
+        AbstractList.checkElementIndex(index, length)
```

Comment:
Let's make a copy of `checkElementIndex` and other checks from `AbstractList` companion as member functions.